pkg/storage/allocator_scorer.go

```diff
@@ -733,6 +733,34 @@ func diversityRemovalScore(
 	return minScore
 }
 
+// diversityRemovalAverageScore is the same as diversityScore, but for a node that's
+// already present in existingNodeLocalities. It works by calculating the
+// average diversityScore for nodeID as if nodeID didn't already have a replica.
+// As with diversityScore, a higher score indicates that the node is a better
+// fit for the range (i.e. keeping it around is good for diversity).
+func diversityRemovalAverageScore(
+	nodeID roachpb.NodeID, existingNodeLocalities map[roachpb.NodeID]roachpb.Locality,
+) float64 {
+	var sumScore float64
+	for otherNodeID, otherLocality := range existingNodeLocalities {
+		if otherNodeID == nodeID {
+			continue
+		}
+		for anotherNodeID, anotherLocality := range existingNodeLocalities {
+			if anotherNodeID == nodeID || anotherNodeID == otherNodeID {
+				continue
+			}
+			newScore := anotherLocality.DiversityScore(otherLocality)
+			sumScore += newScore
+		}
+	}
+	if len(existingNodeLocalities) < 3 {
```

Comment:
Could you explain this way of getting the average? It's pretty obtuse, and seeing the magic numbers here scares me.

Reply:
Sorry.  I'm afraid `num - 1` to be zero, so I need to make sure `len(existingNodeLocalities)` is bigger than three, then we could use the formula below.
```
1.0 / (1.0 + sumScore/float64(num*(num-1)))
```

Reply:
Never returning a score of zero seems bad, since currently we use the entire range `[0,1]` and return the score of 0 quite frequently if all replicas are in the same locality. This is also more complicated than it needs to be. How about we just simplify this method to look more like:

```go
var sumScore float64
var numSamples int
for ... {
  if ... { continue }
  for ... {
    if ... { continue }
    sumScore += ...
    numSamples++
  }
}
if numSamples == 0 {
  return 1.0
}
return sumScore / float64(numSamples)
```

Reply:
Hmm, I think `sumScore / float64(numSamples)` will be the inverse order with `diversityScore`, 
right?

Reply:
I'm not sure what you mean. Can you explain? I could be mistaken, but it certainly looks like `sumScore / float64(numSamples)` will return the average diversity score of all pairs of replicas. That'll be different from the `diversityScore` function on line 703 of this file, but being different from that was intended.

Either way, as @BramGruneir mentioned, adding more diversity score logic without a high-level plan for what the end state should be is just creating more work for ourselves. We should focus on getting that plan before worrying too much about the details of this PR.

Reply:
Sorry, My bad expression.

when `diversityScore` is much bigger, which means the specified Node will be a better fit for the `Range`. But for `averageDiversityScore`, when it's much bigger, it means the specified Node will be a worse fit for the `Range`.

Yeah, I understand we should process the definition of `diversityScore` to avoid confusion, so what we should do is write down all the case we have been involved with `diversityScore`, right?

Reply:
> so what we should do is write down all the case we have been involved with diversityScore, right?

Yeah, the plan should cover all uses of diversity within the allocator.